### DISCUSSION TITLE
http or socks5 proxies support

### DISCUSSION THREAD

**Discussion by cxw620** [hyperium/tonic#1206] [Q&A] [answered]
Hey,

I use tonic to build a grpc client. I need to connect to upstream grpc service with proxy. I searched for help in docs, examples and google but I got nothing about setting a proxy. It seems that tonic currently doesn't support connecting to grpc server with proxy . I'm a new tonic user and I found it hard for me to solve such problem alone and I come here for help.

Does tonic support proxy for connecting to grpc service? If not, any alternative ways? 

BTW:
I read the source code of `tonic::transport`. 
![image](https://user-images.githubusercontent.com/70561268/208613938-d0f4346a-38ac-4648-af7f-a525c9f70f2e.png)
I notice that tonic make use of [hyper](https://docs.rs/hyper/0.14.23/hyper/), and we have [hyper_proxy](https://docs.rs/hyper-proxy/latest/hyper_proxy/) as http proxy connector for hyper based applications as well as [hyper_socks2](https://docs.rs/hyper-socks2/latest/hyper_socks2/#) as socks5 connector. It seems that it's possible to develop a proxy feature for tonic to connect to upstream service.

**Comment by LucioFranco** [comment]
I would take a look at the hyper example in the tonic folder and see how you can use it since as you said the transport just use it under the hood. The transport does not support proxies as its designed to be simple.

**Reply by cxw620** [reply]
More than glad to have your timely reply. 
My other question is, will tonic use system proxy to connect to upstream services? Seems like the answer is no.
I don't know if there is anyone who has ever met with such problem. I'm now trying to figure out any possible solution. If you have any ideas please share with me ^_^

**Comment by cxw620** [comment, marked as answer]
The following example works! 
I have marked all the places that can be modified.
```rust
    use hyper::client::{Client, HttpConnector};
    use hyper::Uri;
    use hyper_socks2::SocksConnector;
    use hyper_openssl::HttpsConnector;
    use openssl::ssl::{SslMethod, SslConnector};
    use tonic::metadata::{AsciiMetadataValue, MetadataValue};
    // ......
    // here is your socks5 proxy addr
    // must be like "socks5://host:port", scheme is required by HttpConnector
    let socks5_addr = "socks5://10.10.2.1:10003";
    let mut connector = HttpConnector::new();
    // "enforce_http(false)" is necessary, without it compiler will not complain but it actually will panic :)
    connector.enforce_http(false);
    // now following the official instruction of crate hyper_socks2
    let proxy = SocksConnector {
        proxy_addr: socks5_addr.parse::<Uri>().unwrap(), 
        auth: None,
        connector,
    };
    // we must delete this, because hyper_socks2 will not help us negotiate HTTP2. You MUST implement it by yourself...
    // let proxy = proxy.with_tls().unwrap(); 
    // negotiate h2
    let mut ssl = SslConnector::builder(SslMethod::tls()).unwrap();
    ssl.set_alpn_protos(b"\x02h2").unwrap();
    let connector = HttpsConnector::with_connector(proxy, ssl).unwrap();
    // ok now we get HttpsConnector! And then we will build the hyper client. The following is easy.
    let client = Client::builder().http2_only(true).build(connector);
    // here is the upstream you are going to connect to. Modify it by yourself.
    let uri = Uri::from_static("https://example.com");
    let mut add_origin = tower::service_fn(|mut req: hyper::Request<tonic::body::BoxBody>| {
        let uri = Uri::builder()
            .scheme(uri.scheme().unwrap().clone())
            .authority(uri.authority().unwrap().clone())
            .path_and_query(req.uri().path_and_query().unwrap().clone())
            .build()
            .unwrap();
        *req.uri_mut() = uri;
        client.request(req)
    });
    // ...
    // `grpc_playurl` is generated by tonic
    let mut final_client = grpc_playurl::play_url_client::PlayUrlClient::with_interceptor(
        add_origin,
        move |mut req: tonic::Request<()>| {
            // non binary example 1
            req.metadata_mut()
                .insert("example-key-1", AsciiMetadataValue::from_static("example-value-static"));
            // non binary example 2
            req.metadata_mut().insert(
                "example-key-2",
                AsciiMetadataValue::try_from(example-value-ref).unwrap(),
            );
            // binary example
            // "gen_metadata_bin" is my own func, it returns type MetadataValue<Binary>
            // macro_rules reference
             /*
                macro_rules! grpc_gen_binary_metadata_value {
                    ($raw_data:expr) => {{
                        let mut buffer = vec![];
                        $raw_data.encode(&mut buffer).unwrap();
                        tonic::metadata::BinaryMetadataValue::from_bytes(&buffer)
                    }};
                }
             */
            req.metadata_mut()
                .insert_bin("example-key-bin", gen_metadata_bin(&client_info));
            Ok(req)
        },
    );
    // `play_view` is generated by tonic
    // `request_content` is your gRPC related struct
    let response = final_client.play_view(request_content).await.unwrap();
    println!("{:?}", response);
```

It's really complex and troublesome, I think. However, this is the only way I have figured out, for I'm new in rust, as well as tonic. I hope someone can make some suggestions for optimizing it... 
BTW, will it be possible for tonic to support adding socks5 proxy directly like the following? It will really save much time.
```rust
let channel = Channel::from_static("http://[::1]:50051").with_proxy_socks5("socks5://127.0.0.1:8888").connect().await.unwrap();
```

**Comment by clouedoc** [comment]
Hello,

I added support for SOCKS5 proxies in my Tonic fork: https://github.com/clouedoc/tonic/commit/2e9e2b33b9e1e0e3ca6c89a699939b9fbcff7ad8

The instructions to use the fork can be found here: https://github.com/hyperium/tonic/issues/970#issuecomment-2597690203
(do note that I ended up putting all my patches in the same branch, `patch-user-agent`. I make no promises regarding backwards compatibility, so I would recommend pinning to a specific commit hash)

Here is an example short:

```rust
        let channel = Channel::from_static(addr)
            // The line below is what you need to add to connect to a SOCKS5 proxy.
            .proxy(Uri::from_static("socks5://XXX.XXX.XXX.XXX:1234"), Some(Auth {
                username: "xxxxxxxxxxxxx".to_string(),
                password: "xxxxxxxxxxxxx".to_string(),
            }))?
            .connect()
            .await?;

        let client = MikuServerClient::with_interceptor(
            channel,
            HeadersInterceptor {
                authentication_header,
            },
        )
            .accept_compressed(CompressionEncoding::Gzip);
```
